### PR TITLE
middleware/proxy: Allow non-HTTP upstreams to be health checked

### PR DESCRIPTION
**Pull request by mfischer-zd**
Allow HTTP health check to be performed against a regular DNS upstream server.

~~TODO: Add tests.~~

### CODE REVIEW

**Comment by mfischer-zd** [comment]
@miekg Need a bit of help with the test design.  Is there a recommended miekg/dns analogue to net/http/httptest?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/pr..." ]
>@miekg Need a bit of help with the test design.  Is there a recommended miekg/dns analogue to net/http/httptest?

Integration tests are put in toplevel test/ dir.

There is a dnsrecorder: middleware/pkg/dnsrecorder

/Miek

--
Miek Gieben

**Comment by mfischer-zd** [comment]
@miekg I'm not sure dnsrecorder is what I'm looking for.  I'd like to amend the `TestHealthCheck()` function in middleware/proxy/upstream_test.go, which gets its host list from the `testPool()` function, which is defined in middleware/proxy/policy_test.go.

So I think I want to instantiate a mock upstream DNS server in `TestMain()` in the policy_test.go file.

**Comment by miekg** [comment]
If you need a DNS server that replies to something the easiest way to do that is just to start a coredns server, but those tests that do that are all in tests/ (mostly to get around cyclic deps and have some sanity with the imports)

**Comment by mfischer-zd** [comment]
Tests added and ready for review.

File: middleware/proxy/upstream.go [hunk 1]
```diff
@@ -15,6 +16,8 @@ import (
 	"github.com/coredns/coredns/middleware/pkg/dnsutil"
 	"github.com/coredns/coredns/middleware/pkg/tls"
 
+	"os"
```
Comment (miekg) [inline middleware/proxy/upstream.go]: nit: this needs to be sorted above.

File: middleware/proxy/upstream.go [hunk 2]
```diff
@@ -229,16 +232,32 @@ func parseBlock(c *caddyfile.Dispenser, u *staticUpstream) error {
...
 		}
-		hostURL := host.Name + port + u.HealthCheck.Path
+
+		hostURL := "http://" + net.JoinHostPort(checkHostName, checkPort) + u.HealthCheck.Path
 		if r, err := http.Get(hostURL); err == nil {
 			io.Copy(ioutil.Discard, r.Body)
 			r.Body.Close()
 			host.Unhealthy = r.StatusCode < 200 || r.StatusCode >= 400
 		} else {
+			fmt.Fprintf(os.Stderr, "Health check probe failed: %v\n", err)
```
Comment (miekg) [inline middleware/proxy/upstream.go]: don't think this adds anything

File: test/proxy_http_health_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"io"
```
Comment (miekg) [inline test/proxy_http_health_test.go]: sorting

File: test/proxy_http_health_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	"github.com/coredns/coredns/middleware/proxy"
+	"github.com/coredns/coredns/middleware/test"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
```
Comment (miekg) [inline test/proxy_http_health_test.go:15]: sorting: std lib first, then coredns, then the rest.

File: test/proxy_http_health_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		t.Fatalf("Could not get CoreDNS authoritative instance: %s", err)
+	}
+
+	proxyAddr, _ := CoreDNSServerPorts(proxyInstance, 0)
+	if proxyAddr == "" {
+		t.Fatalf("Could not get CoreDNS authoritative instance UDP listening port")
+	}
+	defer proxyInstance.Stop()
+
+	log.SetOutput(ioutil.Discard)
```
Comment (miekg) [inline test/proxy_http_health_test.go]: Nit: put this at the top of the function.

**Review by miekg — CHANGES_REQUESTED** [review]
Thanks! Some nits, but looks good.

File: middleware/proxy/upstream.go (hunk 2 above)
Comment (mfischer-zd) [reply middleware/proxy/upstream.go]: If an upstream ends up out of rotation due to a failing health check, having the check failures logged would be useful in performing root cause analysis. 

I couldn't find a more generic log infrastructure, other than the error logging middleware - and this is a warning, not an error.  Can you suggest a better way to deliver this diagnostic information to the user?

**Comment by mfischer-zd** [comment]
Updated imports per feedback.

File: test/proxy_http_health_test.go (hunk 4 above)
Comment (mfischer-zd) [reply test/proxy_http_health_test.go:15]: Fixed.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/pr..." ]
>mfischer-zd commented on this pull request.
>
>
>
>>  		if r, err := http.Get(hostURL); err == nil {
> 			io.Copy(ioutil.Discard, r.Body)
> 			r.Body.Close()
> 			host.Unhealthy = r.StatusCode < 200 || r.StatusCode >= 400
> 		} else {
>+			fmt.Fprintf(os.Stderr, "Health check probe failed: %v\n", err)
>
>If an upstream ends up out of rotation due to a failing health check, having the check failures logged would be useful in performing root cause analysis.
>
>I couldn't find a more generic log infrastructure, other than the error logging middleware - and this is a warning, not an error.  Can you suggest a better way to deliver this diagnostic information to the user?

Ack. Just use `log.Printf` with '[WARNING]'

**Comment by mfischer-zd** [comment]
Updated per feedback.

**Review by miekg — APPROVED** [review]

**Comment by miekg** [comment]
Hmmm, I was a bit to quick :/ But we def. need to fix that data race.

**Comment by mfischer-zd** [comment]
Sorry, which data race?

**Comment by miekg** [comment]
travis just failed, I can reproduce locally
~~~
==================
WARNING: DATA RACE
Read at 0x00c420309658 by goroutine 104:
  github.com/coredns/coredns/middleware/proxy.NewStaticUpstreams.func1.1()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/upstream.go:91 +0x53
  github.com/coredns/coredns/middleware/proxy.(*UpstreamHost).Down()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/proxy.go:73 +0xd6
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).Select()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/upstream.go:288 +0x3e7
  github.com/coredns/coredns/middleware/proxy.Proxy.ServeDNS()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/proxy.go:97 +0x254
  github.com/coredns/coredns/middleware/proxy.(*Proxy).ServeDNS()
      <autogenerated>:12 +0xec
  github.com/coredns/coredns/core/dnsserver.(*Server).ServeDNS()
      /home/miek/g/src/github.com/coredns/coredns/core/dnsserver/server.go:190 +0xab9
  github.com/coredns/coredns/core/dnsserver.(*Server).ServePacket.func1()
      /home/miek/g/src/github.com/coredns/coredns/core/dnsserver/server.go:88 +0xb9
  github.com/miekg/dns.HandlerFunc.ServeDNS()
      /home/miek/g/src/github.com/miekg/dns/server.go:84 +0x51
  github.com/miekg/dns.(*Server).serve()
      /home/miek/g/src/github.com/miekg/dns/server.go:579 +0x80c

Previous write at 0x00c420309658 by goroutine 28:
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).healthCheck()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/upstream.go:253 +0x28f
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).HealthCheckWorker()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/upstream.go:272 +0x84

Goroutine 104 (running) created at:
  github.com/miekg/dns.(*Server).serveUDP()
      /home/miek/g/src/github.com/miekg/dns/server.go:533 +0x38c
  github.com/miekg/dns.(*Server).ActivateAndServe()
      /home/miek/g/src/github.com/miekg/dns/server.go:400 +0x420
  github.com/coredns/coredns/core/dnsserver.(*Server).ServePacket()
      /home/miek/g/src/github.com/coredns/coredns/core/dnsserver/server.go:92 +0x20d
  github.com/mholt/caddy.startServers.func1()
      /home/miek/g/src/github.com/mholt/caddy/caddy.go:672 +0xfc

Goroutine 28 (running) created at:
  github.com/coredns/coredns/middleware/proxy.NewStaticUpstreams()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/upstream.go:109 +0xa5d
  github.com/coredns/coredns/middleware/proxy.setup()
      /home/miek/g/src/github.com/coredns/coredns/middleware/proxy/setup.go:18 +0x72
  github.com/mholt/caddy.executeDirectives()
      /home/miek/g/src/github.com/mholt/caddy/caddy.go:583 +0x794
  github.com/mholt/caddy.ValidateAndExecuteDirectives()
      /home/miek/g/src/github.com/mholt/caddy/caddy.go:533 +0x4e2
  github.com/mholt/caddy.startWithListenerFds()
      /home/miek/g/src/github.com/mholt/caddy/caddy.go:437 +0x7b
  github.com/mholt/caddy.Start()
      /home/miek/g/src/github.com/mholt/caddy/caddy.go:429 +0x14d
  github.com/coredns/coredns/test.CoreDNSServer()
      /home/miek/g/src/github.com/coredns/coredns/test/server.go:26 +0x185
  github.com/coredns/coredns/test.TestProxyWithHTTPCheckOK()
      /home/miek/g/src/github.com/coredns/coredns/test/proxy_http_health_test.go:68 +0x541
  testing.tRunner()
      /home/miek/upstream/go/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestProxyWithHTTPCheckOK (0.01s)
	testing.go:610: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/coredns/coredns/test	10.196s
~~~

**Comment by mfischer-zd** [comment]
Oh, Travis failed for a different reason, I noticed.  Looks like URL.Port() is new for Go 1.8.  Do you want me to port back to 1.7, or make Travis test against 1.8?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/pr..." ]
>Oh, Travis failed for a different reason, I noticed.  Looks like URL.Port() is new for Go 1.8.  Do you want me to port back to 1.7, or make Travis test against 1.8?

Having 1.7 works as well would be nice (should document that). I.e current, and
last version of Go are supported.

**Comment by mfischer-zd** [comment]
Stand by for PR to fix both.

**Comment by mfischer-zd** [comment]
Fix in #603

**Pull request merged**